x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -480,6 +517,70 @@ void validateForUpdate(final String apiKeyId, final Authentication authenticatio
         }
     }
 
+    /**
+     * This method removes remote indices privileges (if defined) from the given user's (i.e. limited-by) role descriptors in two cases:
+     *
+     * <ol>
+     * <li>
+     *     When the provided roles in the request do not define remote indices privileges.
+     *     This means that regardless if user's roles have remote access or not the user choose not to define it for an API key.
+     *     Removing remote indices privileges from the user's role descriptors (in this case) would allow the API key to be used
+     *     to search clusters (using legacy CCS) running a version before {@link RoleDescriptor#VERSION_REMOTE_INDICES}.
+     *  </li>
+     *  <li>
+     *      When we are in a mixed cluster in which some of the nodes do not support remote indices.
+     *      Storing these roles would cause parsing issues on old nodes
+     *      (i.e. nodes running on version before {@link RoleDescriptor#VERSION_REMOTE_INDICES}).
+     * </li>
+     *</ol>
+     */
+    static Set<RoleDescriptor> maybeRemoveRemoteIndicesPrivilegesFromUserRoleDescriptors(
+        final Set<RoleDescriptor> userRoleDescriptors,
+        final Collection<RoleDescriptor> requestedRoleDescriptors,
+        final TransportVersion version,
+        final String... apiKeyIds
+    ) {
+
+        final boolean requestHasRoleDescriptorsWithoutRemoteIndicesPrivileges = requestedRoleDescriptors != null
+            && false == requestedRoleDescriptors.isEmpty()
+            && requestedRoleDescriptors.stream().noneMatch(RoleDescriptor::hasRemoteIndicesPrivileges);
+
+        if (requestHasRoleDescriptorsWithoutRemoteIndicesPrivileges || version.before(RoleDescriptor.VERSION_REMOTE_INDICES)) {
+            return userRoleDescriptors.stream().map(roleDescriptor -> {
+                if (roleDescriptor.hasRemoteIndicesPrivileges()) {
+                    logger.info(
+                        "removed remote indices privileges from role [{}] for API key(s) [{}]",
+                        roleDescriptor.getName(),
+                        String.join(", ", apiKeyIds)
```

Comment:
That's a very good point. Even tough I think having API key ID is useful, your concern is valid.
What do you think if we conditionally log API key IDs (e.g. if a number is less than 10), otherwise we simply log the number of IDs?

Reply:
Agreed that having the IDs is useful here. We could also always log up to the first ten IDs, and the number of IDs, in a log message that makes it clear there is potentially more IDs -- that way large bulk requests still have a few IDs to potentially identify them by. I'm good with either approach.

Reply:
> We could also always log up to the first ten IDs

Even better 👍